src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,162 @@
+
+interface IStatefulActionButtonStateFunctionsAndUtils {
+    /**
+     * Optional argument that could be used for logging purposes in the StatefulActionButton.
+     */
+    name: string;
+    /**
+     * Called when this state is set as the current state.
+     * Called after the onStateExit of the previous state if any.
+     */
+    onStateEntry?: (this: StatefulActionButton) => void;
+    /**
+     * Called when this state has been set as the current state and another is about to get set.
+     * Called before the onStateEntry of the next state if any.
+     */
+    onStateExit?: (this: StatefulActionButton) => void;
+}
+
+/**
+ * Represent a transition from one `IStatefulActionButtonState` to another.
+ */
+export interface IStatefulActionButtonTransition {
+    from: IStatefulActionButtonState;
+    to: IStatefulActionButtonState;
+}
+
+/**
+ * The options format for a StatefulActionButton
+ */
+export interface IStatefulActionButtonOptions {
+    /**
+     * An array containing all states used by the StatefulActionButton instance.
+     * If `switchTo` is called with an unknown state, a warning will be emitted
+     * and the transition will not occur.
+     */
+    states: [IStatefulActionButtonState, ...IStatefulActionButtonState[]];
+    /**
+     * An array containing all the state transitions allowed on the StatefulActionButton instance.
+     * If `switchTo` is called with an illegal transitions , a warning will be emitted
+     * and the transition will not occur.
+     */
+    allowedTransitions?: IStatefulActionButtonTransition[];
+    /**
+     * The initial state to be used by the StatefulActionButton instance.
+     * If nullish or missing from the states array, a warning will be emitted,
+     * no ActionButton will be constructed and the element will be hidden
+     */
+    initalState: IStatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: IStatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initalState;
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
```

Comment:
I think this magic template can be held in a const or a static

Reply:
The text of the warning is specific to this error so no, that's just cumbersome